File: C#/Luhn/Luhn.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Luhn
{
    public static int SumOfDigits(string digitsStr)
    {
        var sum = 0;

        foreach (char digit in digitsStr) 
        {
            sum += (int)Char.GetNumericValue(digit);
        }

        return sum;
    }

    public static char GetLastDigit(int number)
    {
        return number.ToString().Last();
    }

    public static string Reverse(string s)
    {
        char[] charArray = s.ToCharArray();
        Array.Reverse(charArray);
        return new string(charArray);
    }

    public static bool IsValid(string number)
    {
        var whitespaceRemoved = number.Where(c => c != ' ');
        var cleanNumber = whitespaceRemoved.Where(char.IsNumber);

        if (cleanNumber.Count() <= 1 || !whitespaceRemoved.All(char.IsDigit))
        {
            return false;
        }

        string modifiedNumber = "";
        int index = 0;
 
        foreach (char digit in cleanNumber.Reverse())
        {
            if (index % 2 == 0)
            {
                modifiedNumber += digit;
            }
            else
            {
                var result = Int32.Parse(digit.ToString()) * 2;
                if(result > 9) 
                {
                    result -= 9;
                }
                modifiedNumber += GetLastDigit(result );
            }

            index++;
        }

        int digitSum = SumOfDigits(Reverse(modifiedNumber));

        return digitSum % 10 == 0;
    }
}

File: TypeScript/luhn.ts
function sumOfDigits(digitsStr : string) : number {
    let sum = 0;
    
    for(let digit of digitsStr) {
        sum += parseInt(digit);
    };

    return sum;
}

function getLastDigit(n : number) : string {
    return n.toString().slice(-1);
}

function reverse(s : string) : string {
    return s.split("").reverse().join("");
}

function isDigit(n : string) : boolean {
    return Boolean([true, true, true, true, true, true, true, true, true, true][Number(n)]);
}

export function valid(digitString: string): boolean {
    const whitespaceRemoved = digitString.replaceAll(" ", "");
    const cleanNumber = whitespaceRemoved.split("").filter(c => isDigit(c));

    if (cleanNumber.length <= 1 || !whitespaceRemoved.split("").every(c => isDigit(c)))
    {
        return false;
    }

    let modifiedNumber = "";
    let index = 0;

    for(let digit of cleanNumber.reverse())
    {
        if (index % 2 == 0)
        {
            modifiedNumber += digit;
        }
        else
        {
            let result = parseInt(digit) * 2;
            if(result > 9) 
            {
                result -= 9;
            }
            modifiedNumber += getLastDigit(result);
        }

        index++;
    }

    const digitSum = sumOfDigits(reverse(modifiedNumber));

    return digitSum % 10 == 0;
};


// 8569 6195 0383 3437
const numbers = [
    "4539 3195 0343 6467",
    "8273 1232 7352 0569",
    ];


numbers.forEach(number => {
    console.log(`${number} IsValid == ${valid(number)}`);
});